text/0000-forbid-assign-in-call.md

```diff
@@ -0,0 +1,373 @@
+
+```rust
+a = 1;
+foo(());
+
+b = 2;
+let x = Some(());
+
+c = 3;
+let y = (|c| c)(());
+```
+
+You may also put the expression inside parenthesis so the assignment doesn't appear directly inside
+a function call, but this is not a good style we'd recommend.
+
+```rust
+foo((a = 1));
+
+let x = Some((b = 2));
+
+let y = (|c| c)((c = 3));
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Basic rules
+
+1. When an assignment is directly used as an argument in a function or method call, forbid.
+
+    ```rust
+    foo(a = 1);
+    //  ^~~~~ error
+    Some(a = 1);
+    //   ^~~~~ error
+    (|c| c)(a = 1);
+    //      ^~~~~ error
+    bar.unwrap_or(a = 1);
+    //            ^~~~~ error
+    cmp::max(left = 1, right = 2);
+    //       ^~~~~~~~  ^~~~~~~~~ errors
+    ```
+
+2. For simplicity, all kinds of assignments are forbidden.
+
+    ```rust
+    foo(a = 1);
+    //  ^~~~~ error
+    foo(*b = 1);
+    //  ^~~~~~ error
+    foo(c.field = 1);
+    //  ^~~~~~~~~~~ error
+    foo(d.0 = 1);
+    //  ^~~~~~~ error
+    foo(e[0] = 1);
+    //  ^~~~~~~~ error
+    foo(***(f + g).stuff() = 1);
+    //  ^~~~~~~~~~~~~~~~~~~~~~ error
+    foo(concat_idents!(h, i) = 1);
+    //  ^~~~~~~~~~~~~~~~~~~~~~~~ error
+    foo(global::J = 1);
+    //  ^~~~~~~~~~~~~ error
+    ```
+
+3. If the assignment is not directly used, allow.
+
+    ```rust
+    foo((a = 1));
+    // allowed, because the assignment is wrapped inside parenthesis
+    foo({ a = 1 });
+    // allowed, because the assignment is wrapped inside a block
+    foo(unsafe { a = 1 });
+    // allowed, because the assignment is wrapped inside an unsafe block
+    foo(return a = 1);
+    // allowed, the call contains a return statement directly, not an assignment
+    foo(if let Some(x) = y { 1 } else { 2 })
+    // allowed, the `=` is part of the if-let expression, not an assignment
+    ```
+
+4. All other operators including an `=` (e.g. compound assignment) are still allowed.
+
+    ```rust
+    foo(a %= 1);
+    foo(a <= 1);
+    foo(a ..= 1);
+    foo(a >>= 1);
+    ```
+
+5. Assignment expressions appearing elsewhere are allowed.
+
+    ```rust
+    let _ = [a = 1];
+    let _ = [a = 1, a = 2];
+    let _ = [a = 1; 5];
+    //^ arrays
+    let _ = (a = 1,);
+    let _ = (a = 1, a = 2);
+    //^ tuples
+    let _ = || a = 1;
+    //^ closures
+    let _ = a = 1;
+    b = a = 1;
+    c += a = 1;
+    //^ assignments
+    let _ = d[a = 1];
+    d[a = 1] = 0;
+    //^ indexing
+    break a = 1;
+    break 'label a = 1;
+    return a = 1;
+    let _ = yield a = 1;
+    //^ return-like expressions
+    match x { _ => a = 1 };
+    //^ match arm
+    let _ = S { a: a = 1 };
+    //^ struct literal
+    ```
+
+## Rules involving macros
+
+6. A macro variable which evaluates to an assignment expression is allowed.
+
+    ```rust
+    macro_rules! call { ($e:expr) => { foo($e) } }
+    call!(a = 1);
+    // allowed, expands to `foo((a = 1))`.
```

Comment:
This gets into the "are macros tokens or not" question that I never understand.  If the macro used a `tt` instead of an `expr` would it be invalid?  Why did the macro expand to something with parens here?

I think the number of cases involved in the "not in these places" definition [still](https://internals.rust-lang.org/t/pre-rfc-disallow-using-assignment-in-a-function-call-in-rust-2018-disallow-f-a-b/7469/23?u=scottmcm) has me inclined towards the "assignment is just a statement" alternative instead (with some things like "match arms can be any statement", perhaps).

Reply:
@scottmcm Yes if it is expanded via `:tt` instead of `:expr` it would be invalid.

"Something with parenthesis" is already the existing behavior of `:expr`. Compare:

```rust
macro_rules! a {
    (expr: $e:expr) => { 3 * $e };
    (tts: $($t:tt)*) => { 3 * $($t)* };
}

fn main() {
    println!("{}", a!(expr: 7 + 13));   // 60
    println!("{}", a!(tts: 7 + 13));    // 34
}
```